### PR TITLE
Use `regexp_parser` to improve `Style/RedundantRegexp...` cops

### PR DESCRIPTION
**Pull request by owst**
As mentioned in #8593 by @marcandre

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/redundant_regexp_character_class.rb [hunk 1]
```diff
@@ -65,19 +48,54 @@ def on_regexp(node)
...
-            yield match_range(node.loc.begin.end, Regexp.last_match)
+          each_single_element_character_class(node) do |char_class|
+            next unless redundant_single_element_character_class?(node, char_class)
+
+            yield node.loc.begin.adjust(begin_pos: 1 + char_class.ts, end_pos: char_class.te)
+          end
+        end
+
+        def each_single_element_character_class(node)
+          Regexp::Parser.parse(pattern_source(node)).each_expression do |expr|
```
Comment (marcandre) [inline lib/rubocop/cop/style/redundant_regexp_character_class.rb:59]: I think you should use `node.parsed_tree` here...

File: lib/rubocop/cop/style/redundant_regexp_character_class.rb [hunk 2]
```diff
@@ -65,19 +48,54 @@ def on_regexp(node)
...
+            next if expr.type != :set || expr.expressions.size != 1
+            next if expr.negative? || %i[posixclass set].include?(expr.expressions.first.type)
+
+            yield expr
           end
+        rescue Regexp::Scanner::ScannerError
```
Comment (marcandre) [inline lib/rubocop/cop/style/redundant_regexp_character_class.rb:65]: This rescue is no longer needed, the gem was fixed.

File: CHANGELOG.md [hunk 3]
```diff
@@ -44,6 +44,7 @@
 * [#8517](https://github.com/rubocop-hq/rubocop/pull/8517): Make `Style/HashTransformKeys` and `Style/HashTransformValues` aware of `to_h` with block. ([@eugeneius][])
 * [#8529](https://github.com/rubocop-hq/rubocop/pull/8529): Mark `Lint/FrozenStringLiteralComment` as `Safe`, but with unsafe auto-correction. ([@marcandre][])
 * [#8602](https://github.com/rubocop-hq/rubocop/pull/8602): Fix usage of `to_enum(:scan, regexp)` to work on TruffleRuby. ([@jaimerave][])
+* [#8625](https://github.com/rubocop-hq/rubocop/pull/8625): Improve `Style/RedundantRegexpCharacterClass` and `Style/RedundantRegexpEscape` by using `regexp_parser` gem. ([@owst][])
```
Comment (bbatsov) [inline CHANGELOG.md:47]: You'll have to move this under `master`.

File: lib/rubocop/cop/style/redundant_regexp_character_class.rb (hunk 2 above)
Comment (owst) [reply lib/rubocop/cop/style/redundant_regexp_character_class.rb:65]: Hmm, maybe I'm missing something - but if I don't add a rescue here, then I get a failure in this example:
```
  context "with a regexp containing invalid \g escape" do
    it 'does not register an offense' do
      # See https://ruby-doc.org/core-2.7.1/Regexp.html#class-Regexp-label-Subexpression+Calls
      # \g should be \g<name>
      expect_no_offenses('foo = /[a]\g/')
    end
  end
```
```
Failures:

  1) RuboCop::Cop::Style::RedundantRegexpCharacterClass with a regexp containing invalid g escape does not register an offense
     Failure/Error:
                 Regexp::Parser.parse(pattern_source(node)).each_expression do |expr|
                   next if expr.type != :set || expr.expressions.size != 1
                   next if expr.negative? || %i[posixclass set].include?(expr.expressions.first.type)

                   yield expr
                 end

     Regexp::Scanner::ScannerError:
       Scan error at '\g'
     # ./lib/rubocop/cop/style/redundant_regexp_character_class.rb:57:in `each_single_element_character_class'
     # ./lib/rubocop/cop/style/redundant_regexp_character_class.rb:49:in `each_redundant_character_class'
     # ./lib/rubocop/cop/style/redundant_regexp_character_class.rb:33:in `on_regexp'
     # ./lib/rubocop/cop/commissioner.rb:99:in `block (2 levels) in trigger_responding_cops'
     # ./lib/rubocop/cop/commissioner.rb:152:in `with_cop_error_handling'
     # ./lib/rubocop/cop/commissioner.rb:98:in `block in trigger_responding_cops'
     # ./lib/rubocop/cop/commissioner.rb:97:in `each'
     # ./lib/rubocop/cop/commissioner.rb:97:in `trigger_responding_cops'
     # ./lib/rubocop/cop/commissioner.rb:70:in `on_regexp'
     # /Users/owen/code/rubocop-ast/lib/rubocop/ast/traversal.rb:70:in `on_lvasgn'
     # ./lib/rubocop/cop/commissioner.rb:72:in `on_lvasgn'
     # /Users/owen/code/rubocop-ast/lib/rubocop/ast/traversal.rb:14:in `walk'
     # ./lib/rubocop/cop/commissioner.rb:85:in `investigate'
     # ./lib/rubocop/cop/team.rb:152:in `investigate_partial'
     # ./lib/rubocop/cop/team.rb:83:in `investigate'
     # ./lib/rubocop/rspec/cop_helper.rb:54:in `_investigate'
     # ./lib/rubocop/rspec/cop_helper.rb:25:in `inspect_source'
     # ./lib/rubocop/rspec/expect_offense.rb:187:in `expect_no_offenses'
     # ./spec/rubocop/cop/style/redundant_regexp_character_class_spec.rb:68:in `block (3 levels) in <top (required)>'
```

File: lib/rubocop/cop/style/redundant_regexp_character_class.rb (hunk 1 above)
Comment (owst) [reply lib/rubocop/cop/style/redundant_regexp_character_class.rb:59]: Unfortunately `parsed_tree` [doesn't support ](https://github.com/rubocop-hq/rubocop/blob/69700a544bcb15079f177ce5df4f808a61f1949f/lib/rubocop/ext/regexp_node.rb#L19-L20)patterns containing an interpolation. @marcandre I see you added that method - do you recall why you return `nil` for patterns containing interpolations? As it stands, this diff causes a bunch of spec failures:
```diff
diff --git a/lib/rubocop/cop/style/redundant_regexp_character_class.rb b/lib/rubocop/cop/style/redundant_regexp_character_class.rb
index 214a84743..7a3251994 100644
--- a/lib/rubocop/cop/style/redundant_regexp_character_class.rb
+++ b/lib/rubocop/cop/style/redundant_regexp_character_class.rb
@@ -54,7 +54,8 @@ module RuboCop
         end
 
         def each_single_element_character_class(node)
-          Regexp::Parser.parse(pattern_source(node)).each_expression do |expr|
+          # Regexp::Parser.parse(pattern_source(node)).each_expression do |expr|
+          node.parsed_tree.each_expression do |expr|
             next if expr.type != :set || expr.expressions.size != 1
             next if expr.negative? || %i[posixclass set].include?(expr.expressions.first.type)
```
mostly of the form `NoMethodError: undefined method `each_expression' for nil:NilClass`.

N.B. I introduced `pattern_source` (which replaces interpolation and comments with spaces) to remove (as far as the cops are concerned) the "effect" of the interpolation, whilst preserving the indentation/width of the interpolation in the pattern

**Comment by owst** [comment]
Thanks for the review @marcandre - I've pushed a conflict resolution (and move of the CHANGELOG entry) and have hopefully addressed your comments

File: lib/rubocop/cop/style/redundant_regexp_character_class.rb (hunk 1 above)
Comment (owst) [reply lib/rubocop/cop/style/redundant_regexp_character_class.rb:57]: Perhaps we could incorporate the "interpolation-blanking" into `parsed_tree`, or some other mechanism of supporting interpolation?

File: lib/rubocop/cop/style/redundant_regexp_character_class.rb (hunk 2 above)
Comment (marcandre) [reply lib/rubocop/cop/style/redundant_regexp_character_class.rb:63]: Interesting. I'm opened an issue on the gem about this 👍

File: lib/rubocop/cop/style/redundant_regexp_character_class.rb (hunk 1 above)
Comment (marcandre) [reply lib/rubocop/cop/style/redundant_regexp_character_class.rb:57]: I like the idea of blanking the interpolations, makes sense and could be quite useful.
Maybe `parsed_tree` could take a keyword argument `:interpolation` with value `:ignore` or `:blank`?

**Review by bbatsov — APPROVED** [review]

**Comment by bbatsov** [comment]
The changes look good, but your branch has to be rebased on top of the current `master` branch due to merge conflicts.

File: lib/rubocop/cop/style/redundant_regexp_character_class.rb (hunk 2 above)
Comment (owst) [reply lib/rubocop/cop/style/redundant_regexp_character_class.rb:63]: https://github.com/ammar/regexp_parser/issues/65 for the record 👍

File: lib/rubocop/cop/style/redundant_regexp_character_class.rb (hunk 1 above)
Comment (owst) [reply lib/rubocop/cop/style/redundant_regexp_character_class.rb]: I've had a go at this - please let me know what you think @marcandre!

Comment (marcandre) [reply lib/rubocop/cop/style/redundant_regexp_character_class.rb]: That's awesome.
Ideally, the changes to `parsed_tree` would be in a different PR, and with at least a test.
It's not clear to me that replacing the interpolations with spaces has no effect, I thought you were going to simply delete them...
I also don't know why the treatment of the comments.

**Comment by bbatsov** [comment]
@owst Can you rebase and squash, please?

**Comment by owst** [comment]
@bbatsov sorry, I need to deal with @marcandre's comment here https://github.com/rubocop-hq/rubocop/pull/8625#discussion_r486044238 - I'll need to pull out another PR for some of this change. 

For the record, the comment-in-extended-regexp-blanking is (currently) required because `regexp_parser` doesn't allow setting Regexp options when parsing a String pattern rather than Regexp directly. Therefore, `regexp_parser` treats the comment as a literal, which can be flagged by the cops in question e.g. 
```
/foo # [x] <- flagged as redundant brackets if comment is not ignored/x
```

**Comment by jaynetics** [comment]
@owst we can add an option to `regexp_parser` to pass regexp flags / options when parsing a regexp body from a string if that is of help here.

**Comment by owst** [comment]
Hi @jaynetics, thanks for the suggestion! I was looking to do just that, I'll try and pull something together over the weekend and open a PR 👍

**Review by marcandre — APPROVED** [review]
LGTM

**Comment by owst** [comment]
@bbatsov / @marcandre - anything more for me to do on this, or can it be merged now?

**Comment by bbatsov** [comment]
🚀

**Pull request merged**